### DISCUSSION TITLE
Context managers type hinting broken (PyCharm)

### DISCUSSION THREAD

**Discussion by Wouterkoorn** [encode/httpx#2134] [Potential Issue] [answered]
Context managers in httpx use a `TypeVar` to hint the returning of `self`:
![image](https://user-images.githubusercontent.com/54982288/159018159-52aaa006-079b-49bd-b7a7-e697b54e59f1.png)

This does not seem to get picked up by PyCharm:
![image](https://user-images.githubusercontent.com/54982288/159017482-48f1d1e8-98b6-4007-baf2-de848bf5a165.png)

aiohttp directly type hints `__aenter__` (which does work for PyCharm):
![image](https://user-images.githubusercontent.com/54982288/159018020-946b8443-695e-4603-90cd-7ddd87b659d7.png)

I have not tested other IDE's. But PyCharm is the only one I use, and it is really annoying to work with this library because I don't have type hinting, and thus also no autocomplete.

**Comment by arthur-tacca** [comment, marked as answer]
It looks to me like a bug in PyCharm rather than httpx (but I'm not associated with the httpx project so that is just a random person's observation). In any case, as a workaround, you can always do

    async with httpx.AsyncClient() as client:
        assert isinstance(client, httpx.AsyncClient)
        # your code here

which is picked up by PyCharm's type hinting. (In fact it used to be the main way to do type hinting in PyCharm before annotations were added to Python.)

**Reply by Wouterkoorn** [reply]
Yeah, thought it might be. Posted it here, because this is the only library where I have encountered this. I guess I'll also raise it with pycharm then, and see what they think.

The following works as well, but both are no real solutions (ugly and should be unnecessary)
```python
async with httpx.AsyncClient() as client:
    client: httpx.AsyncClient
    # your code here
```

**Reply by florimondmanca** [reply]
Originally this TypeVar trick was introduced so as to allow annotations to work with client subclasses.

It works with mypy, at least:

```python
import httpx

async def main() -> None:
    async with httpx.AsyncClient() as client:
        reveal_type(client)

import asyncio; asyncio.run(main())
```

```console
$ mypy debug/test.py
example.py:5:21: note: Revealed type is "httpx._client.AsyncClient*"
```

So this looks like a PyCharm limitation?

**Reply by Wouterkoorn** [reply]
I think so.
The httpx implementation also follows pep 484 https://peps.python.org/pep-0484/#annotating-instance-and-class-methods

(In python 3.11 we could use the Self typing https://peps.python.org/pep-0673/)
